### PR TITLE
Implement Racing Kings

### PR DESCRIPTION
**Pull request by dignissimus**

### CODE REVIEW

**Review by veloce — COMMENTED** [review]
Sorry for the delay. It looks good but to really ensure validity we need `perft` tests.

You can find them at various places:
- https://github.com/niklasf/chessops/blob/master/src/variant.test.ts#L11
- https://github.com/lichess-org/scalachess/blob/master/src/test/resources/racingkings.perft

File: lib/src/position.dart [hunk 1]
```diff
@@ -1720,6 +1721,140 @@ class ThreeCheck extends Position<ThreeCheck> {
...
+  ///
+  /// Throws a [PositionError] if the [Setup] does not meet basic validity
+  /// requirements.
+  /// Optionnaly pass a `ignoreImpossibleCheck` boolean if you want to skip that
+  /// requirement.
+  factory RacingKings.fromSetup(Setup setup, {bool? ignoreImpossibleCheck}) {
+    final pos = RacingKings(
+      board: setup.board,
+      turn: setup.turn,
+      castles: Castles.fromSetup(setup),
```
Comment (lenguyenthanh) [inline lib/src/position.dart:1807]: RacingKings doesn't support either castling or enpassant, so, I think we should either ignore them, or return errors if the setup has them.

File: lib/src/position.dart [hunk 2]
```diff
@@ -1720,6 +1721,140 @@ class ThreeCheck extends Position<ThreeCheck> {
...
+      turn: turn,
+      unmovedRooks: castles.unmovedRooks,
+      epSquare: _legalEpSquare(),
+      halfmoves: halfmoves,
+      fullmoves: fullmoves,
+    ).fen;
+  }
+
+  @override
+  bool hasInsufficientMaterial(Side side) =>
```
Comment (lenguyenthanh) [inline lib/src/position.dart:1829]: This should always return false, because kings can just reach goal.

File: test/position_test.dart [hunk 3]
```diff
@@ -1268,6 +1268,69 @@ void main() {
...
+    test('game ongoing', () {
+      // While the white king is on the back rank
+      // The black king can reach the back rank this turn
+      final position =
+          RacingKings.fromSetup(Setup.parseFen('1K6/7k/8/8/8/8/8/8 b - - 0 1'));
+      expect(position.isVariantEnd, false);
+      expect(position.variantOutcome, null);
+    });
+
+    test('perft', () {
```
Comment (lenguyenthanh) [inline test/position_test.dart:1308]: Great to see perft tests. My suggestion is to perft file format similar to [scalachess](https://github.com/lichess-org/scalachess/tree/master/src/test/resources) and [shakmaty](https://github.com/niklasf/shakmaty/tree/master/tests). But I think it's for another PR.

File: test/position_test.dart [hunk 4]
```diff
@@ -1268,6 +1268,69 @@ void main() {
...
+      for (final test in [
+        // Start position
+        [
+          '8/8/8/8/8/8/krbnNBRK/qrbnNBRQ w - -',
+          21,
+          421,
+          11264,
```
Comment (lenguyenthanh) [inline test/position_test.dart:1315]: Maybe we should add more depths to our perft tests? In my experience, crazy things can happen if we go deep enough.

File: lib/src/position.dart [hunk 5]
```diff
@@ -1722,6 +1723,149 @@ class ThreeCheck extends Position<ThreeCheck> {
...
+    return false;
+  }
+
+  @override
+  Outcome? get variantOutcome {
+    if (!isVariantEnd) return null;
+    if (whiteInGoal && blackInGoal) return Outcome.draw;
+    // If white is in the goal, check
+    // whether black can reach the goal
+    // And vice-versa
```
Comment (lenguyenthanh) [inline lib/src/position.dart:1803]: This comment is not quite correct. We only do this check if White is in goal, not the other way around.

File: lib/src/position.dart [hunk 6]
```diff
@@ -1722,6 +1723,149 @@ class ThreeCheck extends Position<ThreeCheck> {
...
+    Box<Pockets?>? pockets,
+    Side? turn,
+    Castles? castles,
+    Box<Square?>? epSquare,
+    int? halfmoves,
+    int? fullmoves,
+  }) {
+    return RacingKings(
+      board: board ?? this.board,
+      pockets: pockets != null ? pockets.value : this.pockets,
```
Comment (lenguyenthanh) [inline lib/src/position.dart:1859]: maybe just always null for pockets?

File: lib/src/position.dart [hunk 7]
```diff
@@ -1722,6 +1723,149 @@ class ThreeCheck extends Position<ThreeCheck> {
...
+      turn: turn ?? this.turn,
+      castles: castles ?? this.castles,
```
Comment (lenguyenthanh) [inline lib/src/position.dart:1861]: Also always null for castles.

File: lib/src/position.dart [hunk 8]
```diff
@@ -1722,6 +1723,149 @@ class ThreeCheck extends Position<ThreeCheck> {
...
+      epSquare: epSquare != null ? epSquare.value : this.epSquare,
```
Comment (lenguyenthanh) [inline lib/src/position.dart:1862]: No epSquare either

**Review by lenguyenthanh — COMMENTED** [review]
Some minor comments and suggestions. Other than that it looks good. Great job @dignissimus 💯

File: test/position_test.dart (hunk 4 above)
Comment (veloce) [reply test/position_test.dart:1315]: I used the chessops depths actually. Problem is it takes much more time to run the tests with increased depth. If we separate perft tests from other tests why not.

Comment (lenguyenthanh) [reply test/position_test.dart:1315]: yeah, have another tests for Perft is what Shakmaty and Salachess do. Usually we don't have to run it locally, just run it in CI to make sure.

Comment (veloce) [reply test/position_test.dart:1315]: That can be done in another PR :)

File: test/position_test.dart (hunk 3 above)
Comment (dignissimus) [reply test/position_test.dart:1308]: Moved to https://github.com/lichess-org/dartchess/issues/11

File: test/position_test.dart (hunk 4 above)
Comment (dignissimus) [reply test/position_test.dart:1315]: Resolved with https://github.com/lichess-org/dartchess/issues/11

File: lib/src/position.dart (hunk 5 above)
Comment (dignissimus) [reply lib/src/position.dart:1803]: Resolved in b637550

File: lib/src/position.dart (hunk 7 above)
Comment (dignissimus) [reply lib/src/position.dart:1861]: Resolved in 07884c7

File: lib/src/position.dart (hunk 8 above)
Comment (dignissimus) [reply lib/src/position.dart:1862]: Resolved in 07884c7

File: lib/src/position.dart (hunk 6 above)
Comment (dignissimus) [reply lib/src/position.dart]: Resolved in 07884c7

File: lib/src/position.dart (hunk 7 above)
Comment (dignissimus) [reply lib/src/position.dart:1863]: Good catch 😂

Comment (dignissimus) [reply lib/src/position.dart]: Resolved in cef54ce

File: lib/src/position.dart [hunk 9]
```diff
@@ -1722,6 +1723,151 @@ class ThreeCheck extends Position<ThreeCheck> {
...
+    pos.validate(ignoreImpossibleCheck: ignoreImpossibleCheck);
+    return pos;
+  }
+
+  @override
+  String get fen {
+    return Setup(
+      board: board,
+      turn: turn,
+      unmovedRooks: castles.unmovedRooks,
```
Comment (lenguyenthanh) [inline lib/src/position.dart:1835]: both unmovedRooks & epSquare can be null or empty. Sorry, I missed these before 🙏. And those probably only things left.

Comment (dignissimus) [reply lib/src/position.dart]: Resolved in 068b3e6

**Review by lenguyenthanh — APPROVED** [review]
LGTM, great work @dignissimus!

**Comment by veloce** [comment]
Awesome. Thank you @dignissimus and @lenguyenthanh!

**Pull request merged**